### PR TITLE
Move Providers and ProviderOperations from Azure.ResourceManager.Resources into Azure.ResourceManager.Core-nibhati-5862 

### PR DESCRIPTION
**Pull request by nisha-bhatia**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

**Comment by nisha-bhatia** [comment]
ADO item: https://dev.azure.com/azure-mgmt-ex/DotNET%20Management%20SDK/_boards/board/t/DotNET%20Management%20SDK%20Team/Stories/?workitem=5862

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ApiVersions.cs [hunk 1]
```diff
@@ -20,7 +18,7 @@ namespace Azure.ResourceManager.Core
     /// </summary>
     public class ApiVersions
     {
-        private ProvidersOperations ProviderOperations;
+        private ProviderRestOperations ProviderRestOperations;
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/ApiVersions.cs:21]: given this is a "consumer" of the providers operations I think this should be a providersoperations type.  The only thing that should access the providersrestoperation is the providersoperation and providerscontainer classes

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmClient.cs [hunk 2]
```diff
@@ -139,7 +139,8 @@ public virtual SubscriptionContainer GetSubscriptions()
         /// </summary>
         /// <param name="id"> The id of the resourcegroup </param>
         /// <returns> Resource operations of the resource. </returns>
-        public ResourceGroupOperations GetResourceGroupOperations(ResourceGroupResourceIdentifier id)
+        public ResourceGroupOperations GetResourceGroupOperations(ResourceGroupResourceIdentifier
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmClient.cs:142]: undo this change

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/SubscriptionOperations.cs [hunk 3]
```diff
@@ -94,6 +94,15 @@ public virtual LocationContainer GetLocations()
             return new LocationContainer(this);
         }
 
+        /// <summary>
+        /// Gets the provider container under this subscription.
+        /// </summary>
+        /// <returns> The provider container. </returns>
+        public ProviderContainer GetProviders()
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/SubscriptionOperations.cs:101]: Didn't we walk about adding this at both the TenantOperations and SubscriptionOperations level?

**Review by m-nash — CHANGES_REQUESTED** [review]

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ApiVersions.cs [hunk 4]
```diff
@@ -138,7 +132,7 @@ public string TryGetApiVersion(ResourceType resourceType, CancellationToken canc
             {
                 return val;
             }
-            return ProviderOperations == null ? null : LoadApiVersion(resourceType.ToString(), cancellationToken);
+            return _armClient == null ? null : LoadApiVersion(resourceType.ToString(), cancellationToken); //_armClient is null
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/ApiVersions.cs:135]: what does this comment refer to?

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmClient.cs [hunk 5]
```diff
@@ -82,7 +82,7 @@ public ArmClient(TokenCredential credential, ArmClientOptions options = default)
         /// <param name="baseUri"> The base URI of the service. </param>
         /// <param name="credential"> A credential used to authenticate to an Azure Service. </param>
         /// <param name="options"> The client parameters to use in these operations. </param>
-        private ArmClient(
+        internal ArmClient(
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmClient.cs:85]: I think this can still be private

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderContainer.cs [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        /// <returns></returns>
+        public Response<Provider> Get(string resourceProviderNamespace, CancellationToken cancellationToken = default)
+        {
+            using var scope = Diagnostics.CreateScope("ProviderOperations.Get");
+            scope.Start();
+
+            try
+            {
+                Response<ProviderData> result;
+                if (Id.GetType().Name == "SubscriptionResourceIdentifier")
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderContainer.cs:73]: safer to do an is check here
```suggestion
                if (Id is SubscriptionResourceIdentifier)
```

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderContainer.cs [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+                {
+                    result = RestClient.Get(resourceProviderNamespace, null, cancellationToken);
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderContainer.cs:75]: what is the null parameter here?

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderContainer.cs [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+            {
+                using var scope = Diagnostics.CreateScope("ProvidersContainer.List");
+                scope.Start();
+                string subscriptionId = "";
+
+                try
+                {
+                    Response<ProviderListResult> response;
+                    if (Id.TryGetSubscriptionId(out subscriptionId))
+                        response = RestClient.List(top, expand, cancellationToken);
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderContainer.cs:139]: whenever there is an else block use { }

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderOperations.cs [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+
+namespace Azure.ResourceManager.Core
+{
+    /// <summary> The Providers service client. </summary>
+    public partial class ProviderOperations : ResourceOperationsBase<TenantProviderIdentifier, Provider>
+    {
+        internal ProviderOperations()
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderOperations.cs:18]: this should be the protected constructor for mocking?

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderOperations.cs [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+                    ManagementPipelineBuilder.Build(Credential, BaseUri, ClientOptions),
+                    subscription,
+                    BaseUri);
+            }
+        }
+
+        /// <summary> Unregisters a subscription from a resource provider. </summary>
+        /// <param name="resourceProviderNamespace"> The namespace of the resource provider to unregister. </param>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        public virtual async Task<Response<ProviderData>> UnregisterAsync(string resourceProviderNamespace, CancellationToken cancellationToken = default)
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderOperations.cs:70]: should this return a providerData or a provider?

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/SubscriptionOperations.cs [hunk 11]
```diff
@@ -85,6 +85,15 @@ public virtual LocationContainer GetLocations()
             return new LocationContainer(this);
         }
 
+        /// <summary>
+        /// Gets the provider container under this subscription.
+        /// </summary>
+        /// <returns> The provider container. </returns>
+        public ProviderContainer GetProviders()
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/SubscriptionOperations.cs:92]: make this virtual

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/TenantOperations.cs [hunk 12]
```diff
@@ -65,5 +65,15 @@ public virtual AsyncPageable<T> ListResourcesAsync<T>(Func<Uri, TokenCredential,
         {
             return func(BaseUri, Credential, ClientOptions, Pipeline);
         }
+
+        //TODO:
+        /// <summary>
+        /// Gets the provider container under this subscription.
+        /// </summary>
+        /// <returns> The provider container. </returns>
+        public ProviderContainer GetProviders()
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/TenantOperations.cs:74]: make this virtual

File: sdk/resourcemanager/Azure.ResourceManager.Core/tests/Scenario/ProviderContainerTests.cs [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+            ProviderContainer providerContainer = Client.DefaultSubscription.GetProviders();
+            Assert.Throws<ArgumentNullException>(() => { providerContainer.Get(null); });
+        }
+
+        [TestCase]
+        [RecordedTest]
+        public void GetEmptyException()
+        {
+            ProviderContainer providerContainer = Client.DefaultSubscription.GetProviders();
+            Assert.Throws<ArgumentException>(() => { providerContainer.Get(""); });
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/tests/Scenario/ProviderContainerTests.cs:47]: always use async in tests

File: sdk/resourcemanager/Azure.ResourceManager.Core/tests/Scenario/ProviderContainerTests.cs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+            var result = await providerContainer.GetAsync(resourceNamespace);
+            Assert.IsNotNull(result);
+        }
+
+        [TestCase]
+        [RecordedTest]
+        public void GetNullException()
+        {
+            ProviderContainer providerContainer = Client.DefaultSubscription.GetProviders();
+            Assert.Throws<ArgumentNullException>(() => { providerContainer.Get(null); });
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/tests/Scenario/ProviderContainerTests.cs:39]: always use async in tests

File: sdk/resourcemanager/Azure.ResourceManager.Core/tests/Scenario/ProviderOperationsTests.cs [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+            Response<Provider> response = providerContainer.Get("microsoft.insights");
+            Assert.Throws<ArgumentNullException>(() => {response.Value.Register(null); });
+        }
+
+        [TestCase]
+        [RecordedTest]
+        public void RegisterEmptyException()
+        {
+            ProviderContainer providerContainer = Client.DefaultSubscription.GetProviders();
+            Response<Provider> response = providerContainer.Get("microsoft.insights");
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/tests/Scenario/ProviderOperationsTests.cs:51]: always use async in tests

File: sdk/resourcemanager/Azure.ResourceManager.Core/tests/Scenario/ProviderOperationsTests.cs [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+            var unregister = await result.UnregisterAsync("testerNamespace");
+            Assert.IsNotNull(unregister);
+        }
+
+        [TestCase]
+        [RecordedTest]
+        public void UnregisterNullException()
+        {
+            ProviderContainer providerContainer = Client.DefaultSubscription.GetProviders();
+            Response<Provider> response = providerContainer.Get("microsoft.insights");
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/tests/Scenario/ProviderOperationsTests.cs:71]: always use async in tests

File: sdk/resourcemanager/Azure.ResourceManager.Core/tests/Unit/ArmClientOptionsTests.cs [hunk 17]
```diff
@@ -2,14 +2,17 @@
...
 using NUnit.Framework;
 
 namespace Azure.ResourceManager.Core.Tests
 {
     [Parallelizable]
     public class ArmClientOptionsTests : ResourceManagerTestBase
     {
-        public ArmClientOptionsTests(bool isAsync) : base(isAsync) { }
+        public ArmClientOptionsTests(bool isAsync) : 
+            base(isAsync, RecordedTestMode.Record)
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/tests/Unit/ArmClientOptionsTests.cs:14]: always check in with record mode commented out

**Review by m-nash — CHANGES_REQUESTED** [review]

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderContainer.cs (hunk 7 above)
Comment (nisha-bhatia) [reply sdk/resourcemanager/Azure.ResourceManager.Core/src/Generated/ProviderContainer.cs:75]: the null parameter corresponds to the optional expand parameter, I added it to the Get in ProviderContainer.cs, but I can't add expand to the get in ProviderOperations because the Get in ProviderOperations is an overriden method.

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ApiVersions.cs [hunk 18]
```diff
@@ -106,18 +100,18 @@ private string LoadApiVersion(ResourceType resourceType, CancellationToken cance
             return null;
         }
 
-        private async Task<string> LoadApiVersionAsync(ResourceType resourceType, CancellationToken cancellationToken)
+        private async Task<string> LoadApiVersionAsync(ResourceType resourceType, CancellationToken cancellationToken, string expand = null)
```
Comment (allenjzhang) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/ApiVersions.cs:103]: No need of exposing expand for List/LoadApiVersion.

**Review by allenjzhang — APPROVED** [review]
